runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt

```diff
@@ -29,23 +30,13 @@ import kotlin.coroutines.coroutineContext
  * Otherwise, if it will be sent as a trailing header, calculate the checksum as asynchronously as the body is streamed.
  * In this case, a [LazyAsyncValue] will be added to the execution context which allows the trailing checksum to be sent
  * after the entire body has been streamed.
- *
- * @param checksumAlgorithmNameInitializer a function which parses the input [I] to return the checksum algorithm name
  */
 @InternalApi
-public class FlexibleChecksumsRequestInterceptor<I>(
-    private val checksumAlgorithmNameInitializer: (I) -> String?,
-) : HttpInterceptor {
-    private var checksumAlgorithmName: String? = null
-
-    override fun readAfterSerialization(context: ProtocolRequestInterceptorContext<Any, HttpRequest>) {
-        @Suppress("UNCHECKED_CAST")
-        val input = context.request as I
-        checksumAlgorithmName = checksumAlgorithmNameInitializer(input)
-    }
+public class FlexibleChecksumsRequestInterceptor : HttpInterceptor {
+    override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.logger<FlexibleChecksumsRequestInterceptor>()
 
-    override suspend fun modifyBeforeRetryLoop(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
```

Comment:
This may be leftover from my original interpretation of the SEP's checksum rules (which was wrong), I'll take a look and see if I can revert it.

Reply:
Reminded myself... it's because the endpoint is resolved after `modifyBeforeRetryLoop`. S3 Express needs the checksum set to CRC32 only if `backend=s3express` (endpoint attribute) and that isn't known at the `modifyBeforeRetryLoop` stage. 

So, I had to push back all these checksum-related interceptors to give S3 Express a chance to set CRC32 before flexible checksums / MD5 checksum runs.

Reply:
We spoke offline and agreed this is fine as long as the checksum is cached so it's not recomputed every attempt. 

That's not easy to do for the `deferredChecksum` (chunked) route because that checksum is only done being calculated once the interceptor is done executing. I added caching for the non-chunked request checksums.